test/integration/alpn_selection_integration_test.cc

```diff
@@ -0,0 +1,210 @@
+      auto* transport_socket = cluster->mutable_transport_socket();
+      TestUtility::loadFromYaml(transport_socket_yaml, *transport_socket);
+    });
+    HttpIntegrationTest::initialize();
+  }
+
+  Network::TransportSocketFactoryPtr createUpstreamSslContext() {
+    envoy::extensions::transport_sockets::tls::v3::DownstreamTlsContext tls_context;
+    const std::string yaml = absl::StrFormat(
+        R"EOF(
+common_tls_context:
+  alpn_protocols: [%s]
+  tls_certificates:
+  - certificate_chain: { filename: "%s" }
+    private_key: { filename: "%s" }
+  validation_context:
+    trusted_ca: { filename: "%s" }
+require_client_certificate: true
+)EOF",
+        absl::StrJoin(upstream_alpn_, ","),
+        TestEnvironment::runfilesPath("test/config/integration/certs/upstreamcert.pem"),
+        TestEnvironment::runfilesPath("test/config/integration/certs/upstreamkey.pem"),
+        TestEnvironment::runfilesPath("test/config/integration/certs/cacert.pem"));
+    TestUtility::loadFromYaml(yaml, tls_context);
+    auto cfg = std::make_unique<Extensions::TransportSockets::Tls::ServerContextConfigImpl>(
+        tls_context, factory_context_);
+    static Stats::Scope* upstream_stats_store = new Stats::IsolatedStoreImpl();
+    return std::make_unique<Extensions::TransportSockets::Tls::ServerSslSocketFactory>(
+        std::move(cfg), context_manager_, *upstream_stats_store, std::vector<std::string>{});
+  }
+
+  void createUpstreams() override {
+    auto endpoint = upstream_address_fn_(0);
+    fake_upstreams_.emplace_back(new FakeUpstream(
+        createUpstreamSslContext(), endpoint->ip()->port(),
+        use_h2_ ? FakeHttpConnection::Type::HTTP2 : FakeHttpConnection::Type::HTTP1,
+        endpoint->ip()->version(), timeSystem()));
+  }
+
+  bool use_h2_{};
+  std::vector<std::string> upstream_alpn_;
+  std::vector<std::string> configured_alpn_;
+};
+
+// No upstream ALPN is specified in the protocol, but we successfully negotiate h2 ALPN
+// due to the default ALPN set through the HTTP/2 conn pool.
+TEST_F(AlpnSelectionIntegrationTest, Http2UpstreamMatchingAlpn) {
+  use_h2_ = true;
+  upstream_alpn_.emplace_back(Http::Utility::AlpnNames::get().Http2);
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  IntegrationStreamDecoderPtr response =
+      codec_client_->makeHeaderOnlyRequest(default_request_headers_);
+
+  waitForNextUpstreamRequest();
+  EXPECT_EQ(Http::Utility::AlpnNames::get().Http2,
+            fake_upstream_connection_->connection().nextProtocol());
+
+  upstream_request_->encodeHeaders(default_response_headers_, true);
+  response->waitForEndStream();
+  EXPECT_EQ("200", response->headers().getStatusValue());
+}
+
+// No upstream ALPN is specified in the protocol and we fail to negotiate h2 ALPN
+// since the upstream doesn't list h2 in its ALPN list. Note that the call still goes
+// through because ALPN negotiation failure doesn't necessarily fail the call.
```

Comment:
Should it? This seems kind of broken in practice and unlikely to work, though I suppose we would correctly handle the wrong protocol? I'm wondering though if we should fail this in a more obvious way as a TODO?

Reply:
I think the problem is that the upstream code really doesn't care about what the ALPN negotiation resulted with, we always fall back to the protocol selection policy. This means that ALPN negotiation really only matters on the server side in most cases. 

Looking at the RFC it seems like this should result in a server side rejection: https://tools.ietf.org/html/rfc7301#section-3.2

```
In the event that the server supports no
   protocols that the client advertises, then the server SHALL respond
   with a fatal "no_application_protocol" alert.
```

I would have expected that BoringSSL would handle this on its own, but maybe this is something that we need to do ourselves?

Reply:
AFAIK it's all app level. I would just put some TODOs here as I think we can do better than we do currently. cc @PiotrSikora who has a better understanding of this from a spec perspective than I do.

Reply:
Sounds reasonable, I'll add some TODOs and file an issue